### ISSUE TITLE
Coordinated omission problem

### ISSUE THREAD

**Issue by vjuranek** [PerfCake/PerfCake#258]
It seems, that PerfCake suffers from so called coordinated omission problem (see e.g. [this post](https://groups.google.com/forum/#!msg/mechanical-sympathy/icNZJejUHfE/BfDekfBEs_sJ) for more details what coordinate omission is or one of the numerous presentation on this topic on YT or elsewhere by Gil Tene, usually called "How NOT to Measure Latency").

I tested it only with HTTP, but it's quite likely that the problem is also in other senders, at least those which are not async (as the thread pool is constant). I run default `http` scenario for 30 s. I got stable response time from my server of ~50 ms. In another run, I suspended the server for 15 s after 7 s from beginning of the measurement (i.e. server for responding for ~7 s, then was stuck for 15 s and again responded for ~8 s). Instead of getting average response time of about 7.5 s, I got 86 ms, which is wrong by almost 2 orders of magnitude. Whole output of this run is bellow. 

```
[0:00:01][1872 iterations][3%] [49.254832 ms] [warmUp => false] [Threads => 100] [Minimum => 2.103779 ms] [Maximum => 280.233829 ms] [Average => 49.88392116293391 ms]
[0:00:02][3741 iterations][7%] [74.8074 ms] [warmUp => false] [Threads => 100] [Minimum => 1.846433 ms] [Maximum => 280.233829 ms] [Average => 51.16859593013698 ms]
[0:00:03][5695 iterations][10%] [59.506602 ms] [warmUp => false] [Threads => 100] [Minimum => 1.74227 ms] [Maximum => 280.233829 ms] [Average => 51.380361723172975 ms]
[0:00:04][7740 iterations][13%] [78.750726 ms] [warmUp => false] [Threads => 100] [Minimum => 1.620464 ms] [Maximum => 280.233829 ms] [Average => 50.706216022085584 ms]
[0:00:05][11130 iterations][18%] [56.772474 ms] [warmUp => false] [Threads => 100] [Minimum => 1.620464 ms] [Maximum => 302.028956 ms] [Average => 48.78993410198908 ms]
[0:00:06][12746 iterations][22%] [54.549816 ms] [warmUp => false] [Threads => 100] [Minimum => 1.620464 ms] [Maximum => 551.607736 ms] [Average => 50.502159083787376 ms]
[0:00:07][15228 iterations][25%] [32.669819 ms] [warmUp => false] [Threads => 100] [Minimum => 1.620464 ms] [Maximum => 551.607736 ms] [Average => 48.830329998415074 ms]
[0:00:09][16802 iterations][30%] [2.539824 ms] [warmUp => false] [Threads => 100] [Minimum => 1.620464 ms] [Maximum => 551.607736 ms] [Average => 48.22907156628949 ms]
[0:00:10][16802 iterations][33%] [2.539824 ms] [warmUp => false] [Threads => 100] [Minimum => 1.620464 ms] [Maximum => 551.607736 ms] [Average => 48.22907156628949 ms]
[0:00:11][16802 iterations][37%] [2.539824 ms] [warmUp => false] [Threads => 100] [Minimum => 1.620464 ms] [Maximum => 551.607736 ms] [Average => 48.22907156628949 ms]
[0:00:12][16802 iterations][42%] [2.539824 ms] [warmUp => false] [Threads => 100] [Minimum => 1.620464 ms] [Maximum => 551.607736 ms] [Average => 48.22907156628949 ms]
[0:00:13][16802 iterations][45%] [2.539824 ms] [warmUp => false] [Threads => 100] [Minimum => 1.620464 ms] [Maximum => 551.607736 ms] [Average => 48.22907156628949 ms]
[0:00:14][16802 iterations][48%] [2.539824 ms] [warmUp => false] [Threads => 100] [Minimum => 1.620464 ms] [Maximum => 551.607736 ms] [Average => 48.22907156628949 ms]
[0:00:15][16802 iterations][52%] [2.539824 ms] [warmUp => false] [Threads => 100] [Minimum => 1.620464 ms] [Maximum => 551.607736 ms] [Average => 48.22907156628949 ms]
[0:00:17][16802 iterations][57%] [2.539824 ms] [warmUp => false] [Threads => 100] [Minimum => 1.620464 ms] [Maximum => 551.607736 ms] [Average => 48.22907156628949 ms]
[0:00:18][16802 iterations][62%] [2.539824 ms] [warmUp => false] [Threads => 100] [Minimum => 1.620464 ms] [Maximum => 551.607736 ms] [Average => 48.22907156628949 ms]
[0:00:19][16802 iterations][65%] [2.539824 ms] [warmUp => false] [Threads => 100] [Minimum => 1.620464 ms] [Maximum => 551.607736 ms] [Average => 48.22907156628949 ms]
[0:00:20][16802 iterations][69%] [2.539824 ms] [warmUp => false] [Threads => 100] [Minimum => 1.620464 ms] [Maximum => 551.607736 ms] [Average => 48.22907156628949 ms]
[0:00:21][16802 iterations][72%] [2.539824 ms] [warmUp => false] [Threads => 100] [Minimum => 1.620464 ms] [Maximum => 551.607736 ms] [Average => 48.22907156628949 ms]
[0:00:23][16802 iterations][77%] [2.539824 ms] [warmUp => false] [Threads => 100] [Minimum => 1.620464 ms] [Maximum => 551.607736 ms] [Average => 48.22907156628949 ms]
[0:00:24][20416 iterations][82%] [40.000204 ms] [warmUp => false] [Threads => 100] [Minimum => 1.620464 ms] [Maximum => 15109.523846 ms] [Average => 119.97373587671395 ms]
[0:00:25][23047 iterations][85%] [47.671904 ms] [warmUp => false] [Threads => 100] [Minimum => 1.570762 ms] [Maximum => 15109.523846 ms] [Average => 110.5685385116408 ms]
[0:00:27][26874 iterations][90%] [88.695642 ms] [warmUp => false] [Threads => 100] [Minimum => 1.570762 ms] [Maximum => 15109.523846 ms] [Average => 100.27414137557395 ms]
[0:00:28][29531 iterations][94%] [5.968879 ms] [warmUp => false] [Threads => 100] [Minimum => 1.570762 ms] [Maximum => 15109.523846 ms] [Average => 94.76214215913299 ms]
[0:00:29][32138 iterations][97%] [83.60446 ms] [warmUp => false] [Threads => 100] [Minimum => 1.552704 ms] [Maximum => 15109.523846 ms] [Average => 90.08229741971377 ms]
2015-10-28 21:48:11,754 INFO  {org.perfcake.message.generator.DefaultMessageGenerator} Reached test end.
[0:00:30][34678 iterations][100%] [3.442351 ms] [warmUp => false] [Threads => 100] [Minimum => 1.552704 ms] [Maximum => 15109.523846 ms] [Average => 86.29115656188671 ms]
```

**Comment by pmacik** [comment]
Thanks Vojta to create the issue track for this particular issue.
We are aware of the coordinated omission problem and we are already investigating ways to fix it in PerfCake.
Stay tuned! :)

**Comment by marvec** [comment]
Hello Vojta, could you provide me the scenario and the way to configure the server? It seems strange to me that for the given "dead" period, the reported response time is 2.5ms. It seems like the server responded with an error message instead of hanging the connection. PerfCake simply measures the whole time period, I would like to investigate the results you have observed. Thanks!

**Comment by vjuranek** [comment]
Hi Martin,
sorry for late answer. Here is the scenario:
```xml
<?xml version="1.0" encoding="utf-8"?>
<scenario xmlns="urn:perfcake:scenario:5.0">
   <run type="${perfcake.run.type:time}" value="${perfcake.run.duration:30000}"/>
   <generator class="DefaultMessageGenerator" threads="${perfcake.thread.count:100}"/>
   <sender class="HttpSender">
      <target>http://${server.host}</target>
      <property name="method" value="GET"/>
   </sender>
   <reporting>
      <reporter class="ResponseTimeStatsReporter">
        <destination class="ChartDestination">
          <period type="time" value="1000"/>
          <property name="name" value="Response Time"/>
          <property name="group" value="rt"/>
          <property name="yAxis" value="Response Time [ms]"/>
          <property name="attributes" value="Result,Average"/>
        </destination>
        <destination class="ConsoleDestination">
          <period type="time" value="1000"/>
        </destination>
      </reporter>
   </reporting>
   <messages>
      <message uri="plain.txt"/>
   </messages>
</scenario>
```

Server configuration IMHO doesn't matter, I simply suspended the process and later on resumed (i.e. server didn't reply at all) with
`kill -TSTP 18182; sleep 10; kill -CONT 18182`

Not sure where 2.5 ms came from, I re-run and now there's 40 ms during server pause:
```
[0:00:01][2107 iterations][5%] [52.843199 ms] [warmUp => false] [Threads => 100] [Minimum => 2.478674 ms] [Maximum => 216.115442 ms] [Average => 65.99765234607843 ms]
[0:00:02][4089 iterations][9%] [25.34427 ms] [warmUp => false] [Threads => 100] [Minimum => 1.966306 ms] [Maximum => 216.115442 ms] [Average => 58.86585189710282 ms]
[0:00:03][5996 iterations][12%] [42.931438 ms] [warmUp => false] [Threads => 100] [Minimum => 1.966306 ms] [Maximum => 221.070652 ms] [Average => 56.404037273479545 ms]
[0:00:04][8036 iterations][15%] [35.871207 ms] [warmUp => false] [Threads => 100] [Minimum => 1.850308 ms] [Maximum => 221.070652 ms] [Average => 52.3059169772785 ms]
[0:00:06][11342 iterations][20%] [77.826832 ms] [warmUp => false] [Threads => 100] [Minimum => 1.685334 ms] [Maximum => 221.070652 ms] [Average => 50.26317687748581 ms]
[0:00:07][13611 iterations][24%] [43.405718 ms] [warmUp => false] [Threads => 100] [Minimum => 1.685334 ms] [Maximum => 221.070652 ms] [Average => 49.21421442502401 ms]
[0:00:08][15629 iterations][27%] [42.122652 ms] [warmUp => false] [Threads => 100] [Minimum => 1.66489 ms] [Maximum => 221.070652 ms] [Average => 49.1076200202798 ms]
[0:00:09][18961 iterations][32%] [46.097789 ms] [warmUp => false] [Threads => 100] [Minimum => 1.66489 ms] [Maximum => 221.070652 ms] [Average => 48.357980846133366 ms]
[0:00:10][21121 iterations][35%] [49.568399 ms] [warmUp => false] [Threads => 100] [Minimum => 1.66489 ms] [Maximum => 221.070652 ms] [Average => 48.105084018918035 ms]
[0:00:12][21689 iterations][40%] [40.856015 ms] [warmUp => false] [Threads => 100] [Minimum => 1.647238 ms] [Maximum => 221.070652 ms] [Average => 48.20395570589592 ms]
[0:00:13][21689 iterations][45%] [40.856015 ms] [warmUp => false] [Threads => 100] [Minimum => 1.647238 ms] [Maximum => 221.070652 ms] [Average => 48.20395570589592 ms]
[0:00:14][21689 iterations][49%] [40.856015 ms] [warmUp => false] [Threads => 100] [Minimum => 1.647238 ms] [Maximum => 221.070652 ms] [Average => 48.20395570589592 ms]
[0:00:16][21689 iterations][54%] [40.856015 ms] [warmUp => false] [Threads => 100] [Minimum => 1.647238 ms] [Maximum => 221.070652 ms] [Average => 48.20395570589592 ms]
[0:00:17][21689 iterations][59%] [40.856015 ms] [warmUp => false] [Threads => 100] [Minimum => 1.647238 ms] [Maximum => 221.070652 ms] [Average => 48.20395570589592 ms]
[0:00:18][21689 iterations][62%] [40.856015 ms] [warmUp => false] [Threads => 100] [Minimum => 1.647238 ms] [Maximum => 221.070652 ms] [Average => 48.20395570589592 ms]
[0:00:20][21689 iterations][67%] [40.856015 ms] [warmUp => false] [Threads => 100] [Minimum => 1.647238 ms] [Maximum => 221.070652 ms] [Average => 48.20395570589592 ms]
[0:00:21][22268 iterations][70%] [3.367349 ms] [warmUp => false] [Threads => 100] [Minimum => 1.647238 ms] [Maximum => 10070.002024 ms] [Average => 93.02568272070901 ms]
[0:00:22][25625 iterations][75%] [8.96708 ms] [warmUp => false] [Threads => 100] [Minimum => 1.647238 ms] [Maximum => 10070.002024 ms] [Average => 86.49195038684582 ms]
[0:00:24][29057 iterations][80%] [32.372404 ms] [warmUp => false] [Threads => 100] [Minimum => 1.647238 ms] [Maximum => 10070.002024 ms] [Average => 81.4144854292618 ms]
[0:00:25][31419 iterations][84%] [42.508718 ms] [warmUp => false] [Threads => 100] [Minimum => 1.647238 ms] [Maximum => 10070.002024 ms] [Average => 78.41371348567557 ms]
[0:00:26][33760 iterations][87%] [48.428742 ms] [warmUp => false] [Threads => 100] [Minimum => 1.647238 ms] [Maximum => 10070.002024 ms] [Average => 75.90258166688488 ms]
[0:00:27][35610 iterations][90%] [34.616599 ms] [warmUp => false] [Threads => 100] [Minimum => 1.647238 ms] [Maximum => 10070.002024 ms] [Average => 74.07499603717514 ms]
[0:00:28][38698 iterations][95%] [28.555003 ms] [warmUp => false] [Threads => 100] [Minimum => 1.647238 ms] [Maximum => 10070.002024 ms] [Average => 72.6047358368346 ms]
```

**Comment by marvec** [comment]
Developed an emulator server https://github.com/marvec/broken_http_server

**Comment by marvec** [comment]
And will be investigating shortly

**Comment by marvec** [comment]
Before we actually get to implementing this - I wonder what is the desired behavior in your opinion. First, the scenario has a fixed number of threads/clients - how to handle this? I f I want to continue sending requests at a given rate, I need to be able to dynamically increase the number of threads/clients. Is this correct? And once the server gets back to life, should we decrease the number of clients and get back to the original number?

Another solution would be to simply count each running thread multiple times based on the previous results. E.g. if an average response time is 5ms and suddenly I get 15, I would count it 3 times. What do you think about it?

**Comment by marvec** [comment]
Could we use this? https://github.com/LatencyUtils/LatencyUtils

**Comment by marvec** [comment]
The question also remains whether we suffer from coordinated omission problem - the third solution here http://psy-lob-saw.blogspot.cz/2015/03/fixing-ycsb-coordinated-omission.html suggest using more threads. The number of threads is configurable in PerfCake. Also when using ConstantSpeedGenerator, we can set a large enough pool to cover all the pauses in the application under inspection.

For the really long pauses that can be beaten reasonably by increasing the number of threads, I can only imagine to add estimated artificial results to the measurement. WDYT?

**Comment by marvec** [comment]
I am likely to start with a reporter utilizing https://github.com/HdrHistogram/HdrHistogram

**Comment by marvec** [comment]
This will be solved by adding #267 CoordinatedOmission histogram that uses HdrHistogram.
